rust_src/src/keymap.rs

```diff
@@ -317,6 +319,96 @@ pub fn map_keymap_lisp(function: LispObject, keymap: LispObject, sort_first: boo
     LispObject::constant_nil()
 }
 
+/// Call FUN for every binding in MAP and stop at (and return) the parent.
+/// FUN is called with 4 arguments: FUN (KEY, BINDING, ARGS, DATA).  */
+#[no_mangle]
+pub extern "C" fn map_keymap_internal(
+    map: Lisp_Object,
+    fun: map_keymap_function_t,
+    args: Lisp_Object,
+    data: *const c_void,
+) -> Lisp_Object {
+    let mut tail = LispObject::constant_nil();
+    let map = LispObject::from_raw(map);
+    if let Some(cons) = map.as_cons() {
+        let (car, cdr) = cons.as_tuple();
+        if car.eq_raw(Qkeymap) {
+            tail = cdr
+        } else {
+            tail = map
+        }
+    };
+
```

Comment:
Instead of `loop` you could use `while let Some(tail_cons) = tail.as_cons()`

Reply:
Ah. I didn't know you could do the same thing with `while`. Nice. Will fix. Thanks!